clang/test/CodeGenSYCL/native_cpu_mangling.cpp

```diff
@@ -0,0 +1,18 @@
+// This test ensures the native-cpu device generates the expected kernel names,
+// and that the MS mangler doesn't assert on the code below.
+
+// RUN: %clang_cc1 -triple x86_64-pc-windows-msvc   -aux-triple x86_64-pc-windows-msvc   -I %S/Inputs -fsycl-is-device -fsycl-is-native-cpu -emit-llvm -o - -x c++ %s | FileCheck %s
+// RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -aux-triple x86_64-unknown-linux-gnu -I %S/Inputs -fsycl-is-device -fsycl-is-native-cpu -emit-llvm -o - -x c++ %s | FileCheck %s
+// RUN: %clang_cc1 -triple x86_64-unknown-linux-gnu -aux-triple x86_64-pc-windows-msvc   -I %S/Inputs -fsycl-is-device -fsycl-is-native-cpu -emit-llvm -o - -x c++ %s | FileCheck %s
+// RUN: %clang_cc1 -triple x86_64-pc-windows-msvc   -aux-triple x86_64-unknown-linux-gnu -I %S/Inputs -fsycl-is-device -fsycl-is-native-cpu -emit-llvm -o - -x c++ %s | FileCheck %s
+// Todo: check other cpus
+
+#include "sycl.hpp"
+
+struct name1;
+
+void test(sycl::handler &h) {
+  h.parallel_for_work_group<name1>(sycl::range<1>(2),sycl::range<1>(1), [=](sycl::group<1> G) {});
+}    
+
+// CHECK: void @_ZTS5name1(
```

Comment:
`_ZTS5name1` is an Itanium mangled name. Shouldn't there be a check for a Microsoft mangled name? How does this check pass on Windows?

Reply:
All SYCL targets need to generate the same name for a kernel to enable lookup in different device binaries. All SYCL targets in DPC++ use the Itanium mangling for kernel names.

We still need the MS mangler fix though because when targeting `NativeCPU` to Windows the MS C++ ABI is used (causing the assert), but subsequently `SemaSYCL` generates the unique Itanium-mangled name for each kernel.

Reply:
I'm not sure I'm following. We generate a mangled name for the MS C++ ABI that is never used? If so, wouldn't a better fix be to skip generating such a mangled name in the first place?

Reply:
> I'm not sure I'm following. We generate a mangled name for the MS C++ ABI that is never used? If so, wouldn't a better fix be to skip generating such a mangled name in the first place?

We were considering this but it would probably be more intrusive to query/skip/change the kernel name mangling only for NativeCPU kernels targeting Windows than just handling the `CC_OpenCLKernel` convention which is already in core clang. We wanted to avoid changing mangling in the core clang Windows pipeline just for NativeCPU.

Also, `SemaSYCL` currently always generates a "stable name" (using the Itanium mangling) irrespective of what C++ ABI is used. NativeCPU just selects that existing name - it doesn't create a new name.

Reply:
Just to close out this thread... I've still been struggling to follow these explanations, but as mentioned elsewhere, I won't hold this review up since the change should be harmless, but we'll need to revisit this as part of SYCL upstreaming; the use of `CC_OpenCLKernel` in this context just doesn't make sense.